File: example/fortran/legend_demo/legend_demo.f90
program legend_demo
    !! Demonstration of legend functionality following SOLID principles
    !! Shows legend positioning, labeling, and rendering across all backends
    use fortplot
    use, intrinsic :: iso_fortran_env, only: wp => real64
    implicit none

    call basic_legend_example()
    call positioned_legend_example()
    call multi_function_legend_example()
    call legend_box_styling_example()
    
contains

    subroutine basic_legend_example()
        !! Basic legend usage with labeled plots
        real(wp), dimension(50) :: x, y1, y2
        integer :: i
        
        print *, "=== Basic Legend Example ==="
        
        ! Generate data
        x = [(real(i-1, wp) * 0.2_wp, i=1, 50)]
        y1 = sin(x)
        y2 = cos(x)
        
        call figure(figsize=[8.0_wp, 6.0_wp])
        call title("Basic Legend Demo")
        call xlabel("x")
        call ylabel("y")
        
        ! Add labeled plots
        call add_plot(x, y1, label="sin(x)")
        call add_plot(x, y2, label="cos(x)")
        
        ! Add legend with default position (upper right)
        call legend()
        
        call savefig('output/example/fortran/legend_demo/basic_legend.png')
        call savefig('output/example/fortran/legend_demo/basic_legend.pdf')
        call savefig('output/example/fortran/legend_demo/basic_legend.txt')
        
        print *, "Created: basic_legend.png/pdf/txt"
    end subroutine basic_legend_example

    subroutine positioned_legend_example()
        !! Demonstrate different legend positions
        real(wp), dimension(20) :: x, y1, y2
        integer :: i
        
        print *, "=== Legend Positioning Examples ==="
        
        ! Generate test data
        x = [(real(i, wp), i=1, 20)]
        y1 = x**0.5_wp
        y2 = log(x)
        
        ! Upper left position
        ! NOTE: figsize units are in inches (like matplotlib), not pixels
        call figure(figsize=[6.4_wp, 4.8_wp])
        call title("Legend: Upper Left")
        call add_plot(x, y1, label="√x")
        call add_plot(x, y2, label="ln(x)")
        call legend(position="upper left")
        call savefig('output/example/fortran/legend_demo/legend_upper_left.png')
        call savefig('output/example/fortran/legend_demo/legend_upper_left.pdf')
        call savefig('output/example/fortran/legend_demo/legend_upper_left.txt')
        
        ! Upper right position (default)
        call figure(figsize=[6.4_wp, 4.8_wp])
        call title("Legend: Upper Right")
        call add_plot(x, y1, label="√x")
        call add_plot(x, y2, label="ln(x)")
        call legend(position="upper right")
        call savefig('output/example/fortran/legend_demo/legend_upper_right.png')
        call savefig('output/example/fortran/legend_demo/legend_upper_right.pdf')
        call savefig('output/example/fortran/legend_demo/legend_upper_right.txt')
        
        ! Lower left position
        call figure(figsize=[6.4_wp, 4.8_wp])
        call title("Legend: Lower Left")
        call add_plot(x, y1, label="√x")
        call add_plot(x, y2, label="ln(x)")
        call legend(position="lower left")
        call savefig('output/example/fortran/legend_demo/legend_lower_left.png')
        call savefig('output/example/fortran/legend_demo/legend_lower_left.pdf')
        call savefig('output/example/fortran/legend_demo/legend_lower_left.txt')
        
        ! Lower right position
        call figure(figsize=[6.4_wp, 4.8_wp])
        call title("Legend: Lower Right")
        call add_plot(x, y1, label="√x")
        call add_plot(x, y2, label="ln(x)")
        call legend(position="lower right")
        call savefig('output/example/fortran/legend_demo/legend_lower_right.png')
        call savefig('output/example/fortran/legend_demo/legend_lower_right.pdf')
        call savefig('output/example/fortran/legend_demo/legend_lower_right.txt')
        
        print *, "Created: legend_upper_left/right.png/pdf/txt, legend_lower_left/right.png/pdf/txt"
    end subroutine positioned_legend_example

    subroutine multi_function_legend_example()
        !! Complex legend with multiple mathematical functions
        real(wp), dimension(100) :: x, y1, y2, y3, y4
        integer :: i
        
        print *, "=== Multi-Function Legend Example ==="
        
        ! Generate mathematical functions
        x = [(real(i-1, wp) * 0.1_wp, i=1, 100)]
        y1 = exp(-x/2.0_wp) * cos(x)
        y2 = x * exp(-x/3.0_wp)
        ! Avoid division by zero at x=0: define sin(x)/x = 1 at x=0
        where (x == 0.0_wp)
            y3 = 1.0_wp
        elsewhere
            y3 = sin(x) / x
        end where
        y4 = x**2 * exp(-x)
        
        call figure(figsize=[8.0_wp, 6.0_wp])
        call title("Mathematical Functions with Legend")
        call xlabel("x")
        call ylabel("f(x)")
        
        ! Add multiple labeled functions
        ! Use mathtext with braces for multi-character superscripts
        call add_plot(x, y1, label="e^{-x/2}cos(x)")
        call add_plot(x, y2, label="xe^{-x/3}")
        call add_plot(x, y3, label="sin(x)/x")
        call add_plot(x, y4, label="x^{2}e^{-x}")
        
        ! Add legend
        call legend()
        
        call savefig('output/example/fortran/legend_demo/multi_function_legend.png')
        call savefig('output/example/fortran/legend_demo/multi_function_legend.pdf')
        call savefig('output/example/fortran/legend_demo/multi_function_legend.txt')
        
        print *, "Created: multi_function_legend.png/pdf/txt"
    end subroutine multi_function_legend_example

    subroutine legend_box_styling_example()
        !! Demonstrate legend box styling and multiple entries
        real(wp), dimension(50) :: x, y1, y2, y3, y4
        integer :: i
        
        print *, "=== Legend Box Styling Example ==="
        
        ! Generate test data with multiple functions
        x = [(real(i-1, wp) / 49.0_wp * 2.0_wp * 3.14159_wp, i=1, 50)]
        y1 = sin(x)
        y2 = y1 * 0.5_wp
        y3 = cos(x) * 0.7_wp
        y4 = -y1 * 0.3_wp
        
        call figure(figsize=[8.0_wp, 6.0_wp])
        call title("Legend Box Styling Demo")
        call xlabel("x")
        call ylabel("y")
        
        ! Add multiple plots to test legend box styling
        call add_plot(x, y1, label="sin(x)")
        call add_plot(x, y2, label="0.5 sin(x)")
        call add_plot(x, y3, label="0.7 cos(x)")
        call add_plot(x, y4, label="-0.3 sin(x)")
        
        ! Test legend box with default styling
        call legend()
        call savefig('output/example/fortran/legend_demo/legend_box_default.png')
        call savefig('output/example/fortran/legend_demo/legend_box_default.pdf')
        call savefig('output/example/fortran/legend_demo/legend_box_default.txt')
        
        ! Test legend box in different positions
        call legend(position="upper left")
        call savefig('output/example/fortran/legend_demo/legend_box_upper_left.png')
        call savefig('output/example/fortran/legend_demo/legend_box_upper_left.pdf')
        call savefig('output/example/fortran/legend_demo/legend_box_upper_left.txt')
        
        call legend(position="lower right")
        call savefig('output/example/fortran/legend_demo/legend_box_lower_right.png')
        call savefig('output/example/fortran/legend_demo/legend_box_lower_right.pdf')
        call savefig('output/example/fortran/legend_demo/legend_box_lower_right.txt')
        
        print *, "Created: legend_box_default.png/pdf/txt, legend_box_upper_left.png/pdf/txt, legend_box_lower_right.png/pdf/txt"
    end subroutine legend_box_styling_example

end program legend_demo

File: example/python/legend_demo/legend_demo.py
#!/usr/bin/env python3
"""Demonstration of legend functionality following SOLID principles
Shows legend positioning, labeling, and rendering across all backends - Dual mode: fortplot or matplotlib"""

import sys
import numpy as np

# Dual-mode import: --matplotlib uses matplotlib, default uses fortplot
if "--matplotlib" in sys.argv:
    import matplotlib.pyplot as plt
    backend = "matplotlib"
else:
    import fortplot.fortplot as plt
    backend = "fortplot"

def basic_legend_example():
    """Basic legend usage with labeled plots"""
    print(f"=== Basic Legend Example ({backend}) ===")
    
    # Generate data
    x = np.arange(50) * 0.2
    y1 = np.sin(x)
    y2 = np.cos(x)
    
    plt.figure(figsize=(8, 6))
    plt.title("Basic Legend Demo")
    plt.xlabel("x")
    plt.ylabel("y")
    
    # Add labeled plots
    plt.plot(x, y1, linestyle="-", label="sin(x)")
    plt.plot(x, y2, linestyle="--", label="cos(x)")
    
    # Add legend with default position (upper right)
    plt.legend()
    
    plt.savefig('basic_legend.png')
    plt.savefig('basic_legend.pdf')
    
    # Save TXT for fortplot only
    if backend == "fortplot":
        plt.savefig('basic_legend.txt')
    
    if backend == "matplotlib":
        plt.close()
    
    print("Created: basic_legend.png/pdf" + ("" if backend == "matplotlib" else "/txt"))

def positioned_legend_example():
    """Demonstrate different legend positions"""
    print(f"=== Legend Positioning Examples ({backend}) ===")
    
    # Generate test data
    x = np.arange(1, 21, dtype=float)
    y1 = np.sqrt(x)
    y2 = np.log(x)
    
    # Single legend position demo (upper right - default)
    plt.figure(figsize=(8, 6))
    plt.title("Legend Positioning Demo")
    plt.xlabel("x")
    plt.ylabel("y")
    plt.plot(x, y1, linestyle="-", label="sqrt(x)")
    plt.plot(x, y2, linestyle="--", label="log(x)")
    plt.legend()
    
    plt.savefig('legend_positioning.png')
    plt.savefig('legend_positioning.pdf')
    
    # Save TXT for fortplot only
    if backend == "fortplot":
        plt.savefig('legend_positioning.txt')
    
    if backend == "matplotlib":
        plt.close()
    
    print("Created: legend_positioning.png/pdf" + ("" if backend == "matplotlib" else "/txt"))

def multi_function_legend_example():
    """Complex legend with multiple mathematical functions"""
    print(f"=== Multi-Function Legend Example ({backend}) ===")
    
    # Generate mathematical functions
    x = np.arange(100) * 0.1
    y1 = np.exp(-x/2.0) * np.cos(x)
    y2 = x * np.exp(-x/3.0)
    # Handle division by zero for sin(x)/x
    y3 = np.where(x == 0, 1.0, np.sin(x) / x)
    y4 = x**2 * np.exp(-x)
    
    plt.figure(figsize=(10, 7.5))
    plt.title("Mathematical Functions with Legend")
    plt.xlabel("x")
    plt.ylabel("f(x)")
    
    # Add multiple labeled functions
    plt.plot(x, y1, linestyle="-", label="e^(-x/2)cos(x)")
    plt.plot(x, y2, linestyle="--", label="xe^(-x/3)")
    plt.plot(x, y3, linestyle=":", label="sin(x)/x")
    plt.plot(x, y4, linestyle="-.", label="x^2*exp(-x)")
    
    # Add legend
    plt.legend()
    
    plt.savefig('multi_function_legend.png')
    plt.savefig('multi_function_legend.pdf')
    
    # Save TXT for fortplot only
    if backend == "fortplot":
        plt.savefig('multi_function_legend.txt')
    
    if backend == "matplotlib":
        plt.close()
    
    print("Created: multi_function_legend.png/pdf" + ("" if backend == "matplotlib" else "/txt"))

if __name__ == "__main__":
    basic_legend_example()
    positioned_legend_example()
    multi_function_legend_example()